### PR TITLE
first attempt for `iso_gap_oscar`

### PR DESCRIPTION
**Pull request by ThomasBreuer**
Provide `iso_gap_oscar`:
The idea is to compute a `MapFromFunction` object from a given GAP domain,
with `codomain` value an Oscar object.

Currently GAP's finite fields and cyclotomic fields are supported.
However, finite fields that are extensions of non-prime fields
are currently not supported.
(The straightforward approach would be to compute the minimal polynomial
of a primitive element w.r.t. the prime field, and a corresponding basis.
However, this seems to be not reasonable for GAP's `AlgebraicExtension`s.)

Some more remarks:

- The return value is cached on the GAP side, via a GAP attribute.
  In order to provide this attribute, some GAP code must be executed
  at runtime; I do not know a better way.

- It would be tempting to use GAP's method selection for administrating
  the methods for various kinds of GAP domains.
  This would work if one could be sure that the distinguishing flags
  are always stored in the domains, and if the ranking of the GAP methods
  would be safe.
  Currently there is just one method that calls a Julia function which
  runs through the available cases.

- Perhaps the name `iso_gap_oscar` is not a good choice.
  For the other direction, `ring_iso_oscar_gap` had been changed to
  `iso_oscar_gap` (see #969) because the Julia types of the argument determines
  which kind of isomorphism is intended; this argument does not hold
  for the direction from GAP to Oscar.

- The code around `ring_iso` objects was originally stored in
  `src/Groups/matrices`.
  However, it does not really belong to (matrix) groups, thus it should
  better go to `src/GAP`, as well as its tests.

### CODE REVIEW

**Comment by fingolfin** [comment]
Regarding the missing support for relative extensions: of course this can come later. It might also be possible to map those to the relative extensions of finite fields implemented in Hecke -- though I don't think we support *those* in general in the GAP/Groups part, but I guess we want to support these eventually anyway....? @fieker thoughts?

For reference:
```
julia> F = GF(5,2)
Finite field of degree 2 over F_5

julia> R,x = F[:x]
(Univariate Polynomial Ring in x over Finite field of degree 2 over F_5, x)

julia> f=x^2+x+gen(F)+1; isirreducible(f)
true

julia> F2=FiniteField(f)
(Field extension defined by x^2 + x + o + 1 over Finite field of degree 2 over F_5 , a)
```

Or perhaps Hecke already contains code to turn these relative extensions into absolute extensions of the prime field; then we could use that machinery to compute the missing minimal polynomial...  Paging @thofma @fieker

File: src/GAP/iso_gap_oscar.jl [hunk 1]
```diff
@@ -0,0 +1,192 @@
+# Basically the same as the usual preimage function but without a type check
+# since we don't have elem_type(C) in this case
```
Comment (fingolfin) [inline src/GAP/iso_gap_oscar.jl]: Do you mean this?
```suggestion
# since we don't have elem_type(D) in this case
```

File: src/GAP/iso_gap_oscar.jl [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+# `Nemo.GaloisFmpzField` otherwise)).
+function _iso_gap_oscar_field_finite_prime(F::GAP.GapObj)
+   p = characteristic(F)  # of type `fmpz`
+   if p < fmpz(2)^64
+     p = UInt64(p)
+   end
+   G = GF(p)
+   e = GAPWrap.One(F)
+
+   f(x) = G(fmpz(GAPWrap.IntFFE(x)))
```
Comment (fingolfin) [inline src/GAP/iso_gap_oscar.jl:52]: A tiny optimization is possible:
```suggestion
   function f(x) y=GAPWrap.IntFFE(x); return G(y isa Int ? y : fmpz(y))
```
I actually ended up doing this in a bunch of places, and perhaps it'd be useful to have an abbreviation for this... E.g.
```julia
IntegerUnion(y) = fmpz(y)
IntegerUnion(y::Int) = y
```

**Comment by thofma** [comment]
Continuing your example:
```julia
julia> Hecke.absolute_field(F2[1])
(Finite field of degree 4 over F_5, Morphism from Finite field of degree 4 over F_5 to Field extension defined by x^2 + x + o + 1 over Finite field of degree 2 over F_5 )
```

File: src/GAP/iso_gap_oscar.jl [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+   Basis_G = Vector{elem_type(G)}(undef, d)
+   Basis_G[1] = G(1)
+   for i = 2:d
+      Basis_G[i] = Basis_G[i - 1]*gen(G)
+   end
+
+   function f(x)
+      v = GAPWrap.Coefficients(Basis_F, x)
+      v_int = [fmpz(GAPWrap.IntFFE(v[i])) for i = 1:length(v)]
+      return sum([v_int[i]*Basis_G[i] for i = 1:d])
```
Comment (fingolfin) [inline src/GAP/iso_gap_oscar.jl:106]: We have `length(v) == d`, right? If so, perhaps the following avoids an allocation (or perhaps Julia is already clever enough to do it for us?)
```suggestion
      return sum([fmpz(GAPWrap.IntFFE(v[i])*Basis_G[i] for i = 1:d])
```

But isn't there a constructor for finite field elements which directly takes a coefficient list? That ought to be even more efficient (depending on what `basis(G)` returns... hopefully just what we want... but this doesn't seem to be documented anywhere, in fact `basis(::FinField)` seem undocumented??? @fieker @thofma?

File: src/GAP/iso_gap_oscar.jl [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+   function f(x)
+      GAPWrap.IsCyc(x) || error("$x is not a GAP cyclotomic")
+      denom = GAPWrap.DenominatorCyc(x)
+      n = GAPWrap.Conductor(x)
+      mod(N, n) == 0 || error("$x does not lie in the $N-th cyclotomic field")
+      coeffs = GAP.Globals.CoeffsCyc(x * denom, N)
+      GAP.Globals.ReduceCoeffs(coeffs, cycpol)
+      coeffs = Vector{fmpz}(coeffs)
+      coeffs = coeffs[1:dim]
```
Comment (fingolfin) [inline src/GAP/iso_gap_oscar.jl:135]: Just out of curiosity, why is this line necessary?

I also note that we have essentially the same code now in three places

File: src/GAP/iso_gap_oscar.jl [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+# Use a GAP attribute for caching the mapping.
+# The following must be executed at runtime,
+# the function gets called in Oscar's `__init__`.
+function __init_IsoGapOscar()
+    if ! hasproperty(GAP.Globals, :IsoGapOscar)
+      GAP.evalstr("""
+DeclareAttribute( "IsoGapOscar", IsDomain );
+InstallMethod( IsoGapOscar,
+[ IsDomain ],
+D -> Julia.Oscar._iso_gap_oscar( D ) );
```
Comment (fingolfin) [inline src/GAP/iso_gap_oscar.jl]: Doesn't this also work?
```suggestion
InstallMethod( IsoGapOscar, [ IsDomain ], Julia.Oscar._iso_gap_oscar );
```

File: src/GAP/iso_gap_oscar.jl (hunk 3 above)
Comment (thofma) [reply src/GAP/iso_gap_oscar.jl:106]: Correct. There are no constructors taking a list of coefficients and the basis function is also a crouch :)

File: src/GAP/iso_gap_oscar.jl (hunk 5 above)
Comment (ThomasBreuer) [reply src/GAP/iso_gap_oscar.jl]: With this change, one runs into an error because `Julia.Oscar` is not yet bound at the time when GAP reads the code.

File: src/GAP/iso_gap_oscar.jl (hunk 4 above)
Comment (ThomasBreuer) [reply src/GAP/iso_gap_oscar.jl:135]: `GAP.Globals.ReduceCoeffs` does not remove trailing zeroes, and we need a vector of length `dim`.

Concerning the code duplication: Yes. I had already thought about using the inverse of the `iso_oscar_gap` mapping as `iso_gap_oscar` where this is possible.

**Comment by ThomasBreuer** [comment]
Thanks for the comments.

Concerning the relative extensions, one question is what we actually want. If only a ring isomorphism is needed (the earlier name `ring_iso_oscar_gap` can be interpreted this way) then we are free to choose a suitable representation of the field (or whatever structure) on the other side. However, it might be that users expect to get compatible structures, in the set that also the base fields in GAP and Oscar correspond to each other.

File: src/GAP/iso_gap_oscar.jl (hunk 5 above)
Comment (fingolfin) [reply src/GAP/iso_gap_oscar.jl]: OK, I see. Does this pure Julia code work?
```julia
  GAP.Globals.DeclareAttribute( GAP.Obj("IsoGapOscar"), GAP.Globals.IsDomain );
  GAP.Globals.InstallMethod( GAP.Globals.IsoGapOscar, GAP.Obj([ GAP.Globals.IsDomain ]), _iso_gap_oscar );
```

Comment (fingolfin) [reply src/GAP/iso_gap_oscar.jl]: Anyway, it's also fine to leave it as is, I was mostly wondering

Comment (ThomasBreuer) [reply src/GAP/iso_gap_oscar.jl]: @fingolfin Concerning the idea to use pure Julia code: almost, one has to wrap `_iso_gap_oscar` with `GAP.GapObj`.
Good idea.

**Review by fingolfin — APPROVED** [review]
OK so let's merge it, and then next adjust the GAP->Oscar conversion code to use this isomorphism code

**Comment by ThomasBreuer** [comment]
@fingolfin O.k., I will address the "#HACK" comments from #969 and some others

**Comment by fingolfin** [comment]
@ThomasBreuer how is this coming along? Also hooking up `iso_gap_oscar`
